File: workflow-manager/src/main/java/com/smartcommunity/workflowpoc/services/WorkflowService.java
package com.smartcommunity.workflowpoc.services;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.smartcommunity.workflowpoc.domain.Task;
import com.smartcommunity.workflowpoc.domain.Workflow;

import jakarta.annotation.PostConstruct;

/**
 * Service for managing workflows and processing task completions.
 */
@Service
public class WorkflowService implements TaskCompleteProcessor {

    private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(WorkflowService.class);

    @Autowired
    private MessagingService messagingService;

    @Autowired
    private TaskStoreService taskStoreService;

    private Map<String, Workflow> workflows = new HashMap<>();

    @PostConstruct
    public void init() throws Exception {
        messagingService.init(this);
    }

    /**
     * Starts a new workflow by processing its tasks sequentially.
     * 
     * @param workflow The workflow to start.
     */
    public void startWorkflow(Workflow workflow) {
        workflows.put(workflow.getId(), workflow);
        if (workflow.getTasks() != null && !workflow.getTasks().isEmpty()) {
            doTask(workflow.getTasks().get(0), workflow.getId());
        }
    }

    /**
     * Processes task completion notifications.
     * 
     * @param taskId The ID of the completed task.
     * @param success Whether the task was completed successfully.
     */
    @Override
    public void processTaskCompletion(String taskId, boolean success) {
        //search workflows for taskId
        for (Workflow workflow : workflows.values()) {
            for (int i = 0; i < workflow.getTasks().size(); i++) {
                if (workflow.getTasks().get(i).getId().equals(taskId)) {
                    if (success) {
                        logger.info("Task {} of workflow {} completed successfully", taskId, workflow.getId());
                        //proceed to next task if any
                        if (i + 1 < workflow.getTasks().size()) {
                            try {
                                doTask(workflow.getTasks().get(i + 1), workflow.getId());
                            } catch (Exception e) {
                                e.printStackTrace();
                            }
                        }
                    } else {
                        //mark task as failed
                        workflow.getTasks().get(i).setStatus("FAILED");
                    }
                    return;
                }
            }
        }
    }

    private void doTask(Task task, String workflowId) {
        try {
            task.setWorkflowId(workflowId);
            logger.info("Starting task {} of workflow {}", task.getId(), workflowId);
            task = taskStoreService.initiateTask(task);
            messagingService.sendTask(task);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public Workflow getWorkflow(String id) {
        Workflow workflow = workflows.get(id);
        List<Task> tasks = new LinkedList<>();
        for (Task task : workflows.get(id).getTasks()) {
            if (task.getId() != null) {
                tasks.add(taskStoreService.getTaskById(task.getId()));
            } else {
                tasks.add(task);
            }
        }
        workflow.setTasks(tasks);
        return workflow;

    }
}

File: workflow-manager-dotnet/Services/WorkflowService.cs
using workflow_manager_dotnet.Domain;

namespace workflow_manager_dotnet.Services;

public class WorkflowService : ITaskCompleteProcessor
{
    private readonly ILogger<WorkflowService> _logger;
    private readonly RabbitMQService _messagingService;
    private readonly TaskStoreService _taskStoreService;
    private readonly Dictionary<string, Workflow> _workflows = new();

    public WorkflowService(ILogger<WorkflowService> logger, RabbitMQService messagingService, TaskStoreService taskStoreService)
    {
        _logger = logger;
        _messagingService = messagingService;
        _taskStoreService = taskStoreService;
    }

    public void StartWorkflow(Workflow workflow)
    {
        _workflows[workflow.Id!] = workflow;
        if (workflow.Tasks != null && workflow.Tasks.Any())
        {
            DoTask(workflow.Tasks[0], workflow.Id!);
        }
    }

    public void ProcessTaskCompletion(string taskId, bool success)
    {
        foreach (var workflow in _workflows.Values)
        {
            for (int i = 0; i < workflow.Tasks!.Count; i++)
            {
                if (workflow.Tasks[i].Id == taskId)
                {
                    if (success)
                    {
                        _logger.LogInformation("Task {TaskId} of workflow {WorkflowId} completed successfully", taskId, workflow.Id);
                        if (i + 1 < workflow.Tasks.Count)
                        {
                            DoTask(workflow.Tasks[i + 1], workflow.Id!);
                        }
                    }
                    else
                    {
                        workflow.Tasks[i].Status = "FAILED";
                    }
                    return;
                }
            }
        }
    }

    private void DoTask(WorkflowTask task, string workflowId)
    {
        try
        {
            task.WorkflowId = workflowId;
            _logger.LogInformation("Starting task {TaskId} of workflow {WorkflowId}", task.Id, workflowId);
            task = _taskStoreService.InitiateTask(task);
            _messagingService.SendTask(task);
        }
        catch (Exception e)
        {
            _logger.LogError(e, "Error starting task {TaskId}", task.Id);
        }
    }

    public Workflow GetWorkflow(string id)
    {
        var workflow = _workflows[id];
        var tasks = new List<WorkflowTask>();
        foreach (var task in workflow.Tasks!)
        {
            if (!string.IsNullOrEmpty(task.Id))
            {
                var dbTask = _taskStoreService.GetTaskById(task.Id!);
                if (dbTask != null)
                {
                    tasks.Add(dbTask);
                }
            }
            else
            {
                tasks.Add(task);
            }
        }
        workflow.Tasks = tasks;
        return workflow;
    }
}